sources/api/schnauzer/src/v2/template.rs

```diff
@@ -0,0 +1,159 @@
+                    extension_name,
+                    extension_requirements,
+                )
+            })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Hash)]
+pub struct ExtensionRequirement {
+    pub name: ExtensionName,
+    pub version: ExtensionVersion,
+    pub helpers: Vec<HelperName>,
+}
+
+impl ExtensionRequirement {
+    fn from_template_requirements(
+        extension_name: &ExtensionName,
+        requirements: &TemplateExtensionRequirements,
+    ) -> Self {
+        match requirements {
+            TemplateExtensionRequirements::Version(version) => ExtensionRequirement {
+                name: extension_name.clone(),
+                version: version.clone(),
+                helpers: vec![],
+            },
+            TemplateExtensionRequirements::VersionAndHelpers(extension_requirements) => {
+                ExtensionRequirement {
+                    name: extension_name.clone(),
+                    version: extension_requirements.version.clone(),
+                    helpers: extension_requirements.helpers.clone().unwrap_or_default(),
+                }
+            }
+        }
+    }
+}
+
+impl FromStr for Template {
+    type Err = error::Error;
+
+    fn from_str(input_str: &str) -> Result<Template> {
+        // Potential improvement here would be to use `ariadne` crate to emit more helpful error messages
+        let mut parsed_pairs = TemplateParser::parse(Rule::template, input_str)
+            .context(error::GrammarParseSnafu)?
+            .flatten();
+
+        let frontmatter_pair = parsed_pairs
+            .clone()
+            .find(|p| p.as_rule() == Rule::toml_document)
+            .context(error::TemplateParserLogicSnafu {
+                message: "Template parser did not find frontmatter document.",
+            })?;
+
+        let frontmatter_document: TemplateFrontmatter =
+            toml::from_str(frontmatter_pair.as_str()).context(error::FrontmatterParseSnafu)?;
+
+        let template_body = parsed_pairs
+            .find(|p| p.as_rule() == Rule::body)
+            // The pest grammar ensures that there is exactly one template body.
+            .context(error::TemplateParserLogicSnafu {
+                message: "Template parser did not find template body.",
+            })?
+            .as_str();
+
+        Ok(Self {
+            frontmatter: frontmatter_document,
+            body: template_body.to_string(),
+        })
+    }
+}
+
+pub mod error {
+    use snafu::Snafu;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub))]
+    pub enum Error {
+        #[snafu(display("Error when parsing template grammar: '{}'\n\nThis is usually due to errors in frontmatter TOML formatting.", source))]
+        GrammarParse {
+            source: pest::error::Error<super::Rule>,
+        },
+        #[snafu(display("Error when parsing template frontmatter: '{}'", source))]
+        FrontmatterParse { source: toml::de::Error },
```

Comment:
Nit: Elsewhere in our code we've added a newline between error enum variants, just to make reading through them a bit easier/clearer.